aten/src/ATen/native/cuda/Distributions.cu

```diff
@@ -441,11 +443,77 @@ void uniform_kernel_cuda(TensorIterator& iter, double from_, double to_, Generat
+  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, iter.dtype(), "random_cuda", [&] {
+    using accscalar_t = at::acc_type<scalar_t, true>;
+    if (std::is_same<scalar_t, double>::value || std::is_same<scalar_t, int64_t>::value) {
+      // define lambda to mod with range and add base
+      auto random_func = [range, base] __device__ (uint64_t rand) {
+        return static_cast<scalar_t>(rand % range + base);
+      };
+      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls/2>(iter,
+        gen,
+        [] __device__ (curandStatePhilox4_32_10_t* state) -> ulonglong2 {
+          ulonglong2 ret;
+          uint4 rand = curand4(state);
+          ret.x = (static_cast<uint64_t>(rand.x) << 32) | rand.y;
+          ret.y = (static_cast<uint64_t>(rand.z) << 32) | rand.w;
+          return ret;
+        },
+        random_func);
+    } else {
+      auto random_func = [range, base] __device__ (uint32_t rand) {
+        return static_cast<scalar_t>(rand % static_cast<uint32_t>(range) + static_cast<int32_t>(base));
+      };
+      distribution_nullary_kernel<scalar_t, accscalar_t, curand4_engine_calls>(iter,
+        gen,
+        [] __device__ (curandStatePhilox4_32_10_t* state) { return curand4(state); },
+        random_func);
+    }
+   });
+}
+
 Tensor& uniform_cuda_(Tensor& self, double from, double to, Generator* gen) {
   auto iter = TensorIterator::nullary_op(self);
   uniform_kernel_cuda(*iter, from, to, gen);
   return self;
 }
 
+Tensor& random_cuda_(Tensor& self, Generator* gen) {
+  auto iter = TensorIterator::nullary_op(self);
+  uint64_t range;
+  auto iter_scalar_type = iter->dtype();
+  if (isFloatingType(iter_scalar_type)) {
+    AT_DISPATCH_FLOATING_TYPES_AND_HALF(iter_scalar_type, "random_cuda_range_calc", [&] {
+      range = static_cast<uint64_t>((1ULL << std::numeric_limits<scalar_t>::digits) + 1);
+    });
+  } else {
+    AT_DISPATCH_INTEGRAL_TYPES(iter_scalar_type, "random_cuda_range_calc", [&] {
+      range = static_cast<uint64_t>(std::numeric_limits<scalar_t>::max()) + 1;
+    });
+  }
+  random_kernel_cuda(*iter, range, 0, gen);
+  return self;
+}
+
+Tensor& clamped_random_cuda_(Tensor& self, int64_t from, int64_t to, Generator* gen) {
+  AT_CHECK(from < to, "random_ expects 'from' to be less than 'to', but got from=", from, " >= to=", to);
+  auto iter = TensorIterator::nullary_op(self);
+  AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::Half, iter->dtype(), "range_check", [&] {
+    AT_CHECK(from >= std::numeric_limits<scalar_t>::min() && to <= std::numeric_limits<scalar_t>::max(),
+             "[from, to] is outside the range of the tensor data type: ", toString(iter->dtype()),
+             ". Expected [from, to] to be inside [", std::numeric_limits<scalar_t>::min(),
+             ", ", std::numeric_limits<scalar_t>::max(), "]",
+             " but found [from, to] to be [", from, ", ", to, "]");
+  });
+  auto range = static_cast<uint64_t>(to) - static_cast<uint64_t>(from);
```

Comment:
The additional static casts here look kind of questionable. Are you assuming that to and from are non-negative? I don't see any check to this effect. If to/from could be negative, casting them to unsigned is surprising (it might work out in the end, but I don't want to have to work that out in my head without a comment.)

Reply:
The original code did `max_val - min_val`, without the casts. Now, the promotion rules make my head hurt, but what I think happens is you do the subtraction on `int64_t`, and then convert the (now known to be positive) `int64_t` into a `uint64_t`. So it seems the old code did something different.

Reply:
I goofed here may be in an attempt to fix this bug: https://github.com/pytorch/pytorch/issues/16944. Reverted to what it was before for now. I'll fix that bug separately in a different PR when I come to a conclusion for this: https://github.com/pytorch/pytorch/issues/16944#issuecomment-494239032.